File: src/claude_tts/providers/elevenlabs.py
"""ElevenLabs TTS provider."""

import logging
import os
from typing import Optional

from .base import TTSProvider
from ..types import VoiceGender

logger = logging.getLogger(__name__)


class ElevenLabsProvider(TTSProvider):
    """TTS provider using ElevenLabs API."""
    
    # Default voice IDs for each gender
    DEFAULT_VOICES = {
        "male": "pNInz6obpgDQGcFmaJgB",  # Adam
        "female": "21m00Tcm4TlvDq8ikWAM"  # Rachel
    }
    
    def __init__(self, voice_gender: VoiceGender = "female"):
        """Initialize the ElevenLabs provider."""
        super().__init__(voice_gender)
        self._client: Optional[object] = None
        self._voice_id = self._get_voice_id()
        self._api_key = os.getenv("ELEVENLABS_API_KEY")
        self._available = self._check_availability()
    
    @property
    def name(self) -> str:
        """Return the provider name."""
        return "elevenlabs"
    
    def _get_voice_id(self) -> str:
        """Get the voice ID from environment or use default."""
        env_voice = os.getenv("TTS_ELEVENLABS_VOICE_ID")
        if env_voice:
            return env_voice
        return self.DEFAULT_VOICES.get(self.voice_gender, self.DEFAULT_VOICES["female"])
    
    def _check_availability(self) -> bool:
        """Check if ElevenLabs is available."""
        if not self._api_key:
            logger.debug("ElevenLabs API key not found")
            return False
        
        try:
            from elevenlabs import VoiceSettings
            return True
        except ImportError:
            logger.debug("elevenlabs package not installed")
            return False
        except Exception as e:
            logger.debug(f"ElevenLabs not available: {e}")
            return False
    
    def _get_client(self) -> Optional[object]:
        """Get or create the ElevenLabs client."""
        if self._client is None:
            try:
                from elevenlabs import ElevenLabs
                self._client = ElevenLabs(api_key=self._api_key)
            except Exception as e:
                logger.error(f"Failed to initialize ElevenLabs client: {e}")
                return None
        return self._client
    
    def is_available(self) -> bool:
        """Check if the provider is available."""
        return self._available
    
    def speak(self, text: str) -> bool:
        """
        Speak the given text using ElevenLabs.
        
        Args:
            text: The text to speak
            
        Returns:
            True if successful, False otherwise
        """
        client = self._get_client()
        if not client:
            return False
        
        try:
            from elevenlabs import play, VoiceSettings
            
            # Generate audio
            audio = client.generate(
                text=text,
                voice=self._voice_id,
                model="eleven_monolingual_v1",
                voice_settings=VoiceSettings(
                    stability=0.5,
                    similarity_boost=0.5
                )
            )
            
            # Play the audio
            play(audio)
            return True
            
        except Exception as e:
            logger.error(f"ElevenLabs speak failed: {e}")
            return False

File: src/tts/providers/elevenlabs.ts
import { spawn } from 'child_process';
import { BaseTTSProvider } from './base.js';
import { TTSConfig } from '../types.js';
import { promises as fs } from 'fs';
import { tmpdir } from 'os';
import { join } from 'path';

export class ElevenLabsProvider extends BaseTTSProvider {
  readonly name = 'elevenlabs';
  private apiKey: string;
  
  constructor(config: TTSConfig) {
    super(config);
    this.apiKey = config.elevenLabsApiKey || process.env.ELEVENLABS_API_KEY || '';
  }
  
  async isAvailable(): Promise<boolean> {
    if (!this.apiKey) return false;
    
    try {
      const axios = await import('axios');
      const response = await axios.default.get('https://api.elevenlabs.io/v1/voices', {
        headers: { 'xi-api-key': this.apiKey },
        timeout: 5000
      });
      return response.status === 200;
    } catch {
      return false;
    }
  }
  
  async speak(text: string): Promise<boolean> {
    try {
      const axios = await import('axios');
      const voiceId = this.getVoiceId();
      
      const response = await axios.default.post(
        `https://api.elevenlabs.io/v1/text-to-speech/${voiceId}`,
        {
          text,
          model_id: 'eleven_monolingual_v1',
          voice_settings: {
            stability: 0.5,
            similarity_boost: 0.5
          }
        },
        {
          headers: {
            'Accept': 'audio/mpeg',
            'Content-Type': 'application/json',
            'xi-api-key': this.apiKey
          },
          responseType: 'stream'
        }
      );
      
      // Play the audio stream
      await this.playAudioStream(response.data);
      return true;
    } catch (error) {
      console.error('ElevenLabs provider error:', error);
      return false;
    }
  }
  
  private getVoiceId(): string {
    if (this.config.elevenLabsVoiceId) {
      return this.config.elevenLabsVoiceId;
    }
    
    // Default voices
    const gender = this.config.voiceGender || 'female';
    return gender === 'male' ? 'ErXwobaYiN019PkySvjV' : 'EXAVITQu4vr4xnSDxMaL'; // Antoni : Rachel
  }
  
  private async playAudioStream(audioStream: any): Promise<void> {
    // Save to temp file first for better compatibility
    const tempFile = join(tmpdir(), `stts-${Date.now()}.mp3`);
    
    try {
      // Write stream to temp file
      const chunks: Buffer[] = [];
      for await (const chunk of audioStream) {
        chunks.push(Buffer.from(chunk));
      }
      await fs.writeFile(tempFile, Buffer.concat(chunks));
      
      // Play the temp file
      await this.playAudioFile(tempFile);
      
      // Clean up
      await fs.unlink(tempFile).catch(() => {});
    } catch (error) {
      // Clean up on error
      await fs.unlink(tempFile).catch(() => {});
      throw error;
    }
  }
  
  private async playAudioFile(filePath: string): Promise<void> {
    return new Promise((resolve, reject) => {
      const platform = process.platform;
      let playerCmd: string;
      let playerArgs: string[];
      
      if (platform === 'darwin') { // macOS
        playerCmd = 'afplay';
        playerArgs = [filePath];
      } else if (platform === 'win32') { // Windows
        playerCmd = 'powershell';
        playerArgs = ['-c', `(New-Object Media.SoundPlayer '${filePath}').PlaySync()`];
      } else { // Linux
        playerCmd = 'aplay';
        playerArgs = [filePath];
      }
      
      const player = spawn(playerCmd, playerArgs);
      
      player.on('close', (code: number | null) => {
        if (code === 0) resolve();
        else reject(new Error(`Audio player exited with code ${code}`));
      });
      
      player.on('error', reject);
    });
  }
}